File: tutorials/30_Days_Of_Code/python/d9_Recursion.py
# import sys

def memoize(f):
    cache = {}

    def fact(*args):
        if args in cache:
            return cache[args]
        else:
            cache[args] = f(*args)
            return cache[args]

    return fact

@memoize
def factorial(n):
    if ((n == 0) or (n == 1)):
        return 1
    else:
        return n * factorial(n - 1)


print(factorial(int(input())))

File: tutorials/30_Days_Of_Code/php/d9_Recursion.php
<?php
$_fp = fopen("php://stdin", "r");
/* Enter your code here. Read input from STDIN. Print output to STDOUT */
fscanf($_fp,"%d",$n);
function factorial($x) {
    if ($x <= 1) return 1;
    else return $x * factorial($x - 1);
}
echo factorial($n);
?>
